File: AST Interpreter/Java/Token.java
package Java;

public class Token {
    final TokenType type;
    final String lexeme; // the blob of characters
    final Object literal;
    final int line;

    Token(TokenType type, String lexeme, Object literal, int line) {
        this.type = type;
        this.lexeme = lexeme;
        this.literal = literal;
        this.line = line;
    }

    public String toString() {
        return String.format("%s %s %s", type, lexeme, literal);
    }
}

File: AST Interpreter/C++/Frontend/Token.cpp
#include "Token.h"
#include <sstream>

namespace Frontend
{

    std::string Token::toString() const
    {
        std::ostringstream oss;
        oss << tokenTypeToString(type) << " " << lexeme << " ";

        if (std::holds_alternative<std::nullptr_t>(literal))
        {
            oss << "nil";
        }
        else if (std::holds_alternative<double>(literal))
        {
            oss << std::get<double>(literal);
        }
        else if (std::holds_alternative<std::string>(literal))
        {
            oss << std::get<std::string>(literal);
        }
        else
        {
            oss << "unknown";
        }

        return oss.str();
    }

} // namespace Frontend
